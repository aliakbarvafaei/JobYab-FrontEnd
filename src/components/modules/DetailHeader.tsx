import { Avatar, Button, Grid, IconButton, Typography } from "@mui/material";
import BusinessIcon from "@mui/icons-material/Business";
// import LocationOnIcon from "@mui/icons-material/LocationOn";
import BookmarkBorderOutlinedIcon from "@mui/icons-material/BookmarkBorderOutlined";
import { useRef } from "react";
import { PostType } from "../../constants/types";
import { postBookmark } from "../../services/api";
import { useToast } from "../../contexts/ToastState";
import { addItemOnce } from "../../ts/functions";
import { eachToast } from "../../ts/interfaces";

interface DetailHeaderProps {
  onclick?: () => void;
  haveCompanyDetail?: boolean;
  data?: PostType;
}
const DetailHeader = ({
  onclick,
  haveCompanyDetail = true,
  data,
}: DetailHeaderProps) => {
  const windowWidth = useRef(window.innerWidth);
  const { setToastState } = useToast();

  return (
    <Grid
      className=" sm:mr-1 sm:ml-1 md:mr-10 md:ml-10 xl:ml-3 xl:mr-3 xlmin:mr-20 xlmin:ml-20"
      style={{
        border: "1.5px solid #1976D2",
        height: 200,
        borderRadius: 12,
        marginTop: 10,
        position: "relative",
        backgroundImage:
          "linear-gradient(to left top, #1976D2,#1976D7,#1976D2)",
      }}
    >
      <Grid
        className="flex items-center absolute -bottom-3"
        style={{ right: windowWidth.current > 450 ? 40 : 20 }}
      >
        <Grid>
          <Avatar
            style={{
              width: windowWidth.current > 450 ? 160 : 120,
              height: windowWidth.current > 450 ? 160 : 120,
            }}
          />
        </Grid>
        <Grid>
          <Grid item style={{ marginBottom: 15 }}>
            <Typography
              style={{
                fontSize: 24,
                color: "white",
                marginBottom: windowWidth.current > 450 ? 0 : 50,
              }}
            >
              {data?.title}
            </Typography>
          </Grid>
          {windowWidth.current > 450 ? (
            <>
              <Grid item style={{ display: "flex", alignItems: "center" }}>
                <Grid item>
                  <IconButton disabled>
                    <BusinessIcon style={{ color: "white" }} />
                  </IconButton>
                </Grid>
                <Grid item>
                  <Typography style={{ fontSize: 12, color: "white" }}>
                    {data?.user.company_persian_name}
                  </Typography>
                </Grid>
              </Grid>

              {/* <Grid item style={{ display: "flex", alignItems: "center" }}>
                <Grid item>
                  <IconButton disabled>
                    <LocationOnIcon style={{ color: "white" }} />
                  </IconButton>
                </Grid>
                <Grid item>
                  <Typography style={{ fontSize: 12, color: "white" }}>
                    {`${data?.state.title}, ${data?.city.title}`}
                  </Typography>
                </Grid>
              </Grid> */}
            </>
          ) : null}
        </Grid>
      </Grid>
      <Grid item className="absolute left-0 bottom-3">
        {haveCompanyDetail && (
          <>
            <Button
              variant="contained"
              style={{ marginLeft: 5, background: "white", color: "#1976D2" }}
              onClick={onclick}
            >
              اطلاعات شرکت
            </Button>

            <IconButton style={{ marginLeft: 5, padding: 5 }}>
              <BookmarkBorderOutlinedIcon
                style={{ color: "white" }}
                onClick={() => {
                  postBookmark(data?.id.toString() ?? "")
                    .then((res) => {
                      if (res.status === 201) {
                        setToastState((old: Array<eachToast>) =>
                          addItemOnce(old.slice(), {
                            title: "1",
                            description:
                              "آگهی با موفقیت در نشان شده‌ها قرار داده شد.",
                            key: Math.random(),
                          })
                        );
                      }
                    })
                    .catch((err) => {
                      setToastState((old: Array<eachToast>) =>
                        addItemOnce(old.slice(), {
                          title: "2",
                          description: "مشکلی پیش آمده است.",
                          key: Math.random(),
                        })
                      );
                    });
                }}
              />
            </IconButton>
          </>
        )}
      </Grid>
    </Grid>
  );
};
export default DetailHeader;
