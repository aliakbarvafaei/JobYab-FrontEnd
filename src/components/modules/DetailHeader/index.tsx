import { Button, Grid, IconButton, Typography } from "@mui/material";
import BusinessIcon from "@mui/icons-material/Business";
import BookmarkBorderOutlinedIcon from "@mui/icons-material/BookmarkBorderOutlined";
import { useRef } from "react";
import { PostType } from "../../../constants/types";
import { RemoveBookmark, postBookmark } from "../../../services/api";
import { useToast } from "../../../contexts/ToastState";
import { accessToken, addItemOnce } from "../../../ts/functions";
import { eachToast, statesRedux } from "../../../ts/interfaces";
// import DefaultPicture from "../../assets/images/default.png";
import { useHistory } from "react-router-dom";
import { API_URL } from "../../../config";
import { useDispatch, useSelector } from "react-redux";
import BookmarkIcon from "@mui/icons-material/Bookmark";

interface DetailHeaderProps {
  onclick?: () => void;
  haveCompanyDetail?: boolean;
  data?: PostType;
  updateData?: () => void;
}
const DetailHeader = ({
  onclick,
  haveCompanyDetail = true,
  data,
  updateData,
}: DetailHeaderProps) => {
  const windowWidth = useRef(window.innerWidth);
  const { setToastState } = useToast();
  const history = useHistory();
  const { token } = useSelector((state: statesRedux) => state.userAuth);
  const dispatch = useDispatch();
  
  return (
    <Grid
      className=" sm:mr-1 sm:ml-1 md:mr-10 md:ml-10 xl:ml-3 xl:mr-3 xlmin:mr-20 xlmin:ml-20"
      style={{
        border: "1.5px solid var(--primary)",
        height: 200,
        borderRadius: 12,
        marginTop: 10,
        position: "relative",
        backgroundImage:
          "linear-gradient(to left top,var(--primary),var(--primary),var(--primary)",
      }}
    >
      <Grid
        className="flex items-center absolute -bottom-3"
        style={{ right: windowWidth.current > 450 ? 40 : 20 }}
      >
        <Grid>
          <img
            src={
              data?.user.logo === null
                ? // todo....
                  "../../assets/images/default.png"
                : API_URL.split("api")[0] + (data?.user.logo as string)
            }
            alt=""
            style={{
              borderRadius: "50%",
              width: windowWidth.current > 450 ? 160 : 120,
              height: windowWidth.current > 450 ? 160 : 120,
            }}
          />
        </Grid>
        <Grid>
          <Grid item style={{ marginBottom: 15 }}>
            <Typography
              style={{
                fontSize: 24,
                color: "white",
                marginBottom: windowWidth.current > 450 ? 0 : 50,
              }}
            >
              {data?.title}
            </Typography>
          </Grid>
          {windowWidth.current > 450 ? (
            <>
              <Grid item style={{ display: "flex", alignItems: "center" }}>
                <Grid item>
                  <IconButton disabled>
                    <BusinessIcon style={{ color: "white" }} />
                  </IconButton>
                </Grid>
                <Grid item>
                  <Typography style={{ fontSize: 12, color: "white" }}>
                    {data?.user.company_persian_name}
                  </Typography>
                </Grid>
              </Grid>

              {/* <Grid item style={{ display: "flex", alignItems: "center" }}>
                <Grid item>
                  <IconButton disabled>
                    <LocationOnIcon style={{ color: "white" }} />
                  </IconButton>
                </Grid>
                <Grid item>
                  <Typography style={{ fontSize: 12, color: "white" }}>
                    {`${data?.state.title}, ${data?.city.title}`}
                  </Typography>
                </Grid>
              </Grid> */}
            </>
          ) : null}
        </Grid>
      </Grid>
      <Grid item className="absolute left-0 bottom-3">
        {haveCompanyDetail && (
          <>
            <Button
              variant="contained"
              className="!ml-[5px] !bg-white !text-primary"
              onClick={onclick}
            >
              اطلاعات شرکت
            </Button>

            <IconButton
              style={{ marginLeft: 5, padding: 5 }}
              disabled={!token}
              onClick={() => {
                accessToken(dispatch);
                if (data?.is_bookmark) {
                  RemoveBookmark(data.id || 0)
                    .then((res) => {
                      if (res.status === 204) {
                        setToastState((old: Array<eachToast>) =>
                          addItemOnce(old.slice(), {
                            title: "1",
                            description:
                              "آگهی با موفقیت از لیست نشان شده‌ها حذف شد.",
                            key: Math.random(),
                          })
                        );
                        updateData?.();
                      }
                    })
                    .catch((err) => {
                      if (err.response.status === 401) {
                        setToastState((old: Array<eachToast>) =>
                          addItemOnce(old.slice(), {
                            title: "2",
                            description: "ابتدا باید در وبسایت وارد شوید.",
                            key: Math.random(),
                          })
                        );
                        history.replace("/login");
                      } else {
                        setToastState((old: Array<eachToast>) =>
                          addItemOnce(old.slice(), {
                            title: "2",
                            description: "مشکلی پیش آمده است.",
                            key: Math.random(),
                          })
                        );
                      }
                    });
                } else {
                  postBookmark(data?.id.toString() ?? "")
                    .then((res) => {
                      if (res.status === 201) {
                        setToastState((old: Array<eachToast>) =>
                          addItemOnce(old.slice(), {
                            title: "1",
                            description:
                              "آگهی با موفقیت در نشان شده‌ها قرار داده شد.",
                            key: Math.random(),
                          })
                        );
                        updateData?.();
                      }
                    })
                    .catch((err) => {
                      if (err.response.status === 401) {
                        setToastState((old: Array<eachToast>) =>
                          addItemOnce(old.slice(), {
                            title: "2",
                            description: "ابتدا باید در وبسایت وارد شوید.",
                            key: Math.random(),
                          })
                        );
                        history.replace("/login");
                      } else {
                        setToastState((old: Array<eachToast>) =>
                          addItemOnce(old.slice(), {
                            title: "2",
                            description: "مشکلی پیش آمده است.",
                            key: Math.random(),
                          })
                        );
                      }
                    });
                }
              }}
            >
              {data?.is_bookmark ? (
                <BookmarkIcon className={token ? "text-white" : "text-gray"} />
              ) : (
                <BookmarkBorderOutlinedIcon
                  className={token ? "text-white" : "text-gray"}
                />
              )}{" "}
            </IconButton>
          </>
        )}
      </Grid>
    </Grid>
  );
};
export default DetailHeader;
