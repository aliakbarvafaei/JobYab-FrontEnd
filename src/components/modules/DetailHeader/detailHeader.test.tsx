import { render, fireEvent } from "@testing-library/react";
import DetailHeader from ".";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";


const mockStore = configureStore([]);

describe("DetailHeader Component", () => {
  let store: any;
  let history: any;

  beforeEach(() => {
    store = mockStore({
      userAuth: {
        role: "admin",
        token: "some-token",
      },
    });
    history = createMemoryHistory();
  });
  it("renders correctly", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router history={history}>
          <DetailHeader />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders DetailHeader component without error", () => {
    render(
      <Provider store={store}>
        <Router history={history}>
          <DetailHeader />
        </Router>
      </Provider>
    );
  });

  it("calls onclick function when button is clicked", () => {
    const mockOnClick = jest.fn();
    const { getByText } = render(
      <Provider store={store}>
        <Router history={history}>
          <DetailHeader onclick={mockOnClick} />
        </Router>
      </Provider>
    );

    fireEvent.click(getByText("اطلاعات شرکت"));
    expect(mockOnClick).toHaveBeenCalled();
  });
  test('clicking "اطلاعات شرکت" button calls onclick function', () => {
    const mockOnClick = jest.fn();
    const { getByText } = render(
      <Provider store={store}>
        <Router history={history}>
          <DetailHeader onclick={mockOnClick} haveCompanyDetail />
        </Router>
      </Provider>
    );

    const infoButton = getByText("اطلاعات شرکت");
    fireEvent.click(infoButton);

    expect(mockOnClick).toHaveBeenCalled();
  });
});
