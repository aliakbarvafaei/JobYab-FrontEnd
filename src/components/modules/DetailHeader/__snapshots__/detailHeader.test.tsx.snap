// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailHeader Component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root  sm:mr-1 sm:ml-1 md:mr-10 md:ml-10 xl:ml-3 xl:mr-3 xlmin:mr-20 xlmin:ml-20 css-vj1n65-MuiGrid-root"
    style="height: 200px; border-radius: 12px; margin-top: 10px; position: relative;"
  >
    <div
      class="MuiGrid-root flex items-center absolute -bottom-3 css-vj1n65-MuiGrid-root"
      style="right: 40px;"
    >
      <div
        class="MuiGrid-root css-vj1n65-MuiGrid-root"
      >
        <img
          alt=""
          src="http://localhost:8000/undefined"
          style="border-radius: 50%; width: 160px; height: 160px;"
        />
      </div>
      <div
        class="MuiGrid-root css-vj1n65-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          style="margin-bottom: 15px;"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            style="font-size: 24px; color: white; margin-bottom: 0px;"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          style="display: flex; align-items: center;"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <button
              class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="BusinessIcon"
                focusable="false"
                style="color: white;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 7V3H2v18h20V7H12zM6 19H4v-2h2v2zm0-4H4v-2h2v2zm0-4H4V9h2v2zm0-4H4V5h2v2zm4 12H8v-2h2v2zm0-4H8v-2h2v2zm0-4H8V9h2v2zm0-4H8V5h2v2zm10 12h-8v-2h2v-2h-2v-2h2v-2h-2V9h8v10zm-2-8h-2v2h2v-2zm0 4h-2v2h2v-2z"
                />
              </svg>
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              style="font-size: 12px; color: white;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item absolute left-0 bottom-3 css-13i4rnv-MuiGrid-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium !ml-[5px] !bg-white !text-primary css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        اطلاعات شرکت
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        style="margin-left: 5px; padding: 5px;"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium text-white css-i4bv87-MuiSvgIcon-root"
          data-testid="BookmarkBorderOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17 3H7c-1.1 0-2 .9-2 2v16l7-3 7 3V5c0-1.1-.9-2-2-2zm0 15-5-2.18L7 18V5h10v13z"
          />
        </svg>
         
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
